### PR TITLE
Physics: set link velocity from *VelocityReset components

### PR DESCRIPTION
**Pull request by yaswanth1701**
# 🎉 New feature

Closes #<NUMBER>

## Summary

Follow-up PR to #2440.  Added link velocity reset component for linear and angular velocity. Updated the physics system plugin to support the reset components of link velocity.
## Test it

Have to add test case.
## Checklist
- [X] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] Consider updating Python bindings (if the library has them)
- [ ] `codecheck` passed (See [contributing](https://gazebosim.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://gazebosim.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Agazebosim+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge** and edit the commit message to match the pull request summary while retaining `Signed-off-by` messages.

### CODE REVIEW

File: src/systems/physics/Physics.cc [hunk 1]
```diff
@@ -3628,58 +3775,34 @@ void PhysicsPrivate::UpdateSim(EntityComponentManager &_ecm,
...
+        _vel->Data() = math::Vector3d::Zero;
+        return true;
+      });
 
-    for (const auto entity : entitiesLinearVelocityCmd)
-    {
-      _ecm.RemoveComponent<components::LinearVelocityCmd>(entity);
-    }
-  }
-  GZ_PROFILE_END();
```
Comment (scpeters) [inline src/systems/physics/Physics.cc:3682]: this appears to reverse some changes from https://github.com/gazebosim/gz-sim/pull/2228; can you revert that so the `*VelocityCmd` handling is unchanged?

**Review by scpeters — COMMENTED** [review]
I've read the code and left some comments; I'll make a suggestion shortly about how to test this

**Comment by scpeters** [comment]
for a test, I suggest a similar structure to the test for `JointVelocityReset` in https://github.com/gazebosim/gz-sim/blob/gz-sim8/test/integration/physics_system.cc#L750-L847

if we borrow a 1x4x9 box from the boxes benchmark with the "complex" case (gravity and gyroscopic tumbling) we could set the initial linear velocity and expect it to change according to gravity, and we could set the initial angular velocity and expect it to rotate while preserving angular momentum in the world frame

File: src/systems/physics/Physics.cc (hunk 1 above)
Comment (yaswanth1701) [reply src/systems/physics/Physics.cc]: yeah, I didn't notice this sorry for that. I have reverted and made the required changes.

**Comment by yaswanth1701** [comment]
> for a test, I suggest a similar structure to the test for `JointVelocityReset` in https://github.com/gazebosim/gz-sim/blob/gz-sim8/test/integration/physics_system.cc#L750-L847
> 
> if we borrow a 1x4x9 box from the boxes benchmark with the "complex" case (gravity and gyroscopic tumbling) we could set the initial linear velocity and expect it to change according to gravity, and we could set the initial angular velocity and expect it to rotate while preserving angular momentum in the world frame

okay, I will check this out and will add the test.

File: include/gz/sim/components/AngularVelocityReset.hh [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+#ifndef GZ_SIM_COMPONENTS_WORLDANGULARVELOCITYRESET_HH_
+#define GZ_SIM_COMPONENTS_WORLDANGULARELOCITYRESET_HH_
```
Comment (scpeters) [inline include/gz/sim/components/AngularVelocityReset.hh:18]:
```suggestion
#define GZ_SIM_COMPONENTS_WORLDANGULARVELOCITYRESET_HH_
```

**Comment by scpeters** [comment]
I've made an example test for WorldLinearVelocityReset in https://github.com/yaswanth1701/gz-sim/pull/1

please merge that in and then add one for WorldAngularVelocityReset if you have time

**Comment by yaswanth1701** [comment]
> I've made an example test for WorldLinearVelocityReset in [yaswanth1701#1](https://github.com/yaswanth1701/gz-sim/pull/1)
> 
> please merge that in and then add one for WorldAngularVelocityReset if you have time

yeah, sure I will add the test for angular velocity reset.

File: include/gz/sim/components/LinearVelocityReset.hh [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace gz
+{
+namespace sim
+{
+// Inline bracket to help doxygen filtering.
+inline namespace GZ_SIM_VERSION_NAMESPACE {
+namespace components
+{
+  /// \brief Linear velocity of an entity in it's own frame
```
Comment (scpeters) [inline include/gz/sim/components/LinearVelocityReset.hh:34]:
```suggestion
  /// \brief Linear velocity of an entity in its own frame
```

File: test/integration/physics_system.cc [hunk 4]
```diff
@@ -747,6 +749,272 @@ TEST_F(PhysicsSystemFixture,
...
+  }
+
+  const math::Pose3d pose0(0, 0, 1, 0, 0, 0);
+  // To avoid collision with ground plane while rotating
+  model.SetWorldPoseCmd(*ecm, pose0);
+  server.Run(true, 1, false);
+
+  const math::Vector3d vel0(0.1, 5.0, 0.1);
+
+  // Reset child link velocity to nonzero components (upward in Z)
```
Comment (scpeters) [inline test/integration/physics_system.cc:964]:
```suggestion
  // Reset child link velocity to nonzero components
```

File: test/integration/physics_system.cc [hunk 5]
```diff
@@ -747,6 +749,272 @@ TEST_F(PhysicsSystemFixture,
...
+  {
+    auto rootLinkAngularVel = rootLink.WorldAngularVelocity(*ecm);
+    auto childLinkAngularVel = childLink.WorldAngularVelocity(*ecm);
+    ASSERT_NE(std::nullopt, rootLinkAngularVel);
+    ASSERT_NE(std::nullopt, childLinkAngularVel);
+    EXPECT_EQ(math::Vector3d::Zero, rootLinkAngularVel);
+    EXPECT_EQ(math::Vector3d::Zero, childLinkAngularVel);
+  }
+
+  // Reset root link velocity to nonzero components (upward in Z)
```
Comment (scpeters) [inline test/integration/physics_system.cc:980]:
```suggestion
  // Reset root link velocity to nonzero components
```

File: include/gz/sim/components/AngularVelocityReset.hh (hunk 2 above)
Comment (scpeters) [reply include/gz/sim/components/AngularVelocityReset.hh]: fixed in e6607268e76d17765f83847513e6fa53760e0666

**Review by scpeters — COMMENTED** [review]
just a few small suggestions; the new test looks good!

**Comment by yaswanth1701** [comment]
> just a few small suggestions; the new test looks good!

done

**Review by scpeters — APPROVED** [review]

**Pull request merged**